File: iOS/MyAIAgent/MyAIAgent/Networking/ClaudeProvider.swift
//
//  ClaudeProvider.swift
//  MyAIAgent
//

import Foundation

/// Streams the Anthropic Messages API over SSE and maps wire payloads to
/// provider-neutral `StreamEvent`s. The API key arrives via
/// Secrets.xcconfig -> Info.plist -> Bundle; it is never hardcoded.
nonisolated struct ClaudeProvider: ModelProvider {
    static let model = "claude-opus-4-8"
    static let maxTokens = 4096  // short conversational turns; keeps demo cost bounded

    let apiKey: String

    /// Reads the key the build injected. Nil when Secrets.xcconfig is
    /// missing or the env var was unset — the app gates on this at launch.
    static func bundledAPIKey(bundle: Bundle = .main) -> String? {
        guard let key = bundle.object(forInfoDictionaryKey: "AnthropicAPIKey") as? String,
              !key.isEmpty else { return nil }
        return key
    }

    func stream(
        system: String,
        tools: [ToolDefinition],
        messages: [WireMessage]
    ) -> AsyncThrowingStream<StreamEvent, any Error> {
        AsyncThrowingStream { continuation in
            let task = Task {
                do {
                    var request = URLRequest(url: URL(string: "https://api.anthropic.com/v1/messages")!)
                    request.httpMethod = "POST"
                    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
                    request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
                    request.setValue("2023-06-01", forHTTPHeaderField: "anthropic-version")
                    request.httpBody = try JSONEncoder().encode(MessagesRequest(
                        model: Self.model,
                        maxTokens: Self.maxTokens,
                        system: system,
                        tools: tools,
                        messages: messages,
                        stream: true
                    ))

                    let (bytes, response) = try await URLSession.shared.bytes(for: request)
                    if let http = response as? HTTPURLResponse, http.statusCode != 200 {
                        throw http.statusCode >= 500 ? AgentError.server : AgentError.generic
                    }

                    // NOT bytes.lines — it swallows the blank lines that
                    // delimit SSE events (see SSELineSplitter).
                    var splitter = SSELineSplitter()
                    var parser = SSEParser()
                    var toolIDsByIndex: [Int: String] = [:]

                    func process(line: String) throws {
                        guard let event = parser.consume(line: line) else { return }
                        let payload = AnthropicStreamPayload.decode(Data(event.data.utf8))
                        if case .apiError = payload { throw AgentError.server }
                        for mapped in Self.map(payload, toolIDsByIndex: &toolIDsByIndex) {
                            continuation.yield(mapped)
                        }
                    }

                    for try await byte in bytes {
                        if let line = splitter.consume(byte: byte) {
                            try process(line: line)
                        }
                    }
                    if let tail = splitter.flush() {
                        try process(line: tail)
                    }
                    try process(line: "")  // flush any event pending at EOF
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    /// Wire payload -> provider-neutral events. Pure (given the index map)
    /// so the regression suite can pin the mapping without networking.
    static func map(
        _ payload: AnthropicStreamPayload,
        toolIDsByIndex: inout [Int: String]
    ) -> [StreamEvent] {
        switch payload {
        case .contentBlockStartText:
            return [.textStarted]
        case .contentBlockStartToolUse(let index, let id, let name):
            toolIDsByIndex[index] = id
            return [.toolUseStarted(id: id, name: name)]
        case .textDelta(_, let text):
            return [.textDelta(text)]
        case .inputJSONDelta(let index, let partialJSON):
            guard let id = toolIDsByIndex[index] else { return [] }
            return [.toolInputDelta(id: id, partialJSON: partialJSON)]
        case .contentBlockStop(let index):
            guard let id = toolIDsByIndex.removeValue(forKey: index) else { return [] }
            return [.toolUseFinished(id: id)]
        case .messageDelta(let stopReason):
            guard let stopReason else { return [] }
            let reason: StreamEvent.StopReason = switch stopReason {
            case "end_turn": .endTurn
            case "tool_use": .toolUse
            case "max_tokens": .maxTokens
            case let other: .other(other)
            }
            return [.finished(stopReason: reason)]
        case .messageStop, .ignored, .apiError:
            return []
        }
    }
}

File: android/app/src/main/java/com/curiousbeagle/android/myaiagent/net/ClaudeProvider.kt
package com.curiousbeagle.android.myaiagent.net

import com.curiousbeagle.android.myaiagent.model.AgentError
import com.curiousbeagle.android.myaiagent.model.AgentFailure
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.serialization.encodeToString
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.TimeUnit

/**
 * Streams the Anthropic Messages API over SSE and maps wire payloads to
 * provider-neutral [StreamEvent]s. Mirrors iOS `ClaudeProvider`. The API key
 * arrives via local.properties/env -> BuildConfig; it is never hardcoded.
 */
class ClaudeProvider(private val apiKey: String) : ModelProvider {

    companion object {
        const val MODEL = "claude-opus-4-8"
        const val MAX_TOKENS = 4096 // short conversational turns; keeps demo cost bounded
        private const val ENDPOINT = "https://api.anthropic.com/v1/messages"

        /**
         * Wire payload -> provider-neutral events. Pure (given the index
         * map) so the regression suite can pin the mapping — mirrors
         * `ClaudeProvider.map` on iOS.
         */
        fun map(
            payload: AnthropicStreamPayload,
            toolIdsByIndex: MutableMap<Int, String>,
        ): List<StreamEvent> = when (payload) {
            is AnthropicStreamPayload.ContentBlockStartText -> listOf(StreamEvent.TextStarted)
            is AnthropicStreamPayload.ContentBlockStartToolUse -> {
                toolIdsByIndex[payload.index] = payload.id
                listOf(StreamEvent.ToolUseStarted(payload.id, payload.name))
            }
            is AnthropicStreamPayload.TextDelta -> listOf(StreamEvent.TextDelta(payload.text))
            is AnthropicStreamPayload.InputJsonDelta ->
                toolIdsByIndex[payload.index]
                    ?.let { listOf(StreamEvent.ToolInputDelta(it, payload.partialJson)) }
                    ?: emptyList()
            is AnthropicStreamPayload.ContentBlockStop ->
                toolIdsByIndex.remove(payload.index)
                    ?.let { listOf(StreamEvent.ToolUseFinished(it)) }
                    ?: emptyList()
            is AnthropicStreamPayload.MessageDelta -> when (payload.stopReason) {
                null -> emptyList()
                "end_turn" -> listOf(StreamEvent.Finished(StreamEvent.StopReason.EndTurn))
                "tool_use" -> listOf(StreamEvent.Finished(StreamEvent.StopReason.ToolUse))
                "max_tokens" -> listOf(StreamEvent.Finished(StreamEvent.StopReason.MaxTokens))
                else -> listOf(StreamEvent.Finished(StreamEvent.StopReason.Other(payload.stopReason)))
            }
            AnthropicStreamPayload.MessageStop,
            AnthropicStreamPayload.Ignored,
            is AnthropicStreamPayload.ApiError -> emptyList()
        }
    }

    private val client = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(0, TimeUnit.MILLISECONDS) // SSE: the stream stays open between events
        .build()

    override fun stream(
        system: String,
        tools: List<ToolDefinition>,
        messages: List<WireMessage>,
    ): Flow<StreamEvent> = flow {
        val body = wireJson.encodeToString(
            MessagesRequest(MODEL, MAX_TOKENS, system, tools, messages, stream = true)
        )
        val request = Request.Builder()
            .url(ENDPOINT)
            .header("Content-Type", "application/json")
            .header("x-api-key", apiKey)
            .header("anthropic-version", "2023-06-01")
            .post(body.toRequestBody("application/json".toMediaType()))
            .build()

        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                throw AgentFailure(if (response.code >= 500) AgentError.SERVER else AgentError.GENERIC)
            }
            val source = response.body?.source() ?: throw AgentFailure(AgentError.SERVER)

            // readUtf8Line() preserves empty lines — the SSE event delimiter.
            // (URLSession's line convenience on iOS swallowed them; the
            // regression suite pins the parser's dependence on empties.)
            val parser = SseParser()
            val toolIdsByIndex = mutableMapOf<Int, String>()
            while (true) {
                val line = source.readUtf8Line() ?: break
                val event = parser.consume(line) ?: continue
                val payload = AnthropicStreamPayload.decode(event.data)
                if (payload is AnthropicStreamPayload.ApiError) throw AgentFailure(AgentError.SERVER)
                map(payload, toolIdsByIndex).forEach { emit(it) }
            }
            // Flush any event pending at EOF (stream ended without a final blank line).
            parser.consume("")?.let { event ->
                val payload = AnthropicStreamPayload.decode(event.data)
                if (payload is AnthropicStreamPayload.ApiError) throw AgentFailure(AgentError.SERVER)
                map(payload, toolIdsByIndex).forEach { emit(it) }
            }
        }
    }.flowOn(Dispatchers.IO)
}
